reading/queue.livemd

```diff
@@ -0,0 +1,315 @@
+# Queues
+
+```elixir
+Mix.install([
+  {:youtube, github: "brooklinjazz/youtube"},
+  {:hidden_cell, github: "brooklinjazz/hidden_cell"},
+  {:tested_cell, github: "brooklinjazz/tested_cell"},
+  {:utils, path: "#{__DIR__}/../utils"}
+])
+```
+
+## Navigation
+
+[Return Home](../start.livemd)<span style="padding: 0 30px"></span>
+[Report An Issue](https://github.com/DockYard-Academy/beta_curriculum/issues/new?assignees=&labels=&template=issue.md&title=)
+
+## Setup
+
+Ensure you type the `ea` keyboard shortcut to evaluate all Elixir cells before starting. Alternatively, you can evaluate the Elixir cells as you read.
+
+## Queues
+
+Queues are an abstract data structure based on the common human
+experience of waiting in lines. In human lines, people arrive at
+the end of the line and are served from the front of the line.
+
+<!-- livebook:{"break_markdown":true} -->
+
+```mermaid
+flowchart
+  subgraph Queue
+    direction LR
+    Last[Last/Newest] --> NextN[NextN] -- ... --> Next[Next] --> First[First/Oldest]
+  end
+```
+
+<!-- livebook:{"break_markdown":true} -->
+
+In programming we have defined a Queue as:
+
+* The first item in the queue is called the *front* or *first*.
+* The *front* of the queue is gotten and removed when we
+  *dequeue* the queue.
+* The last item in the queue is called the *back* or *last*.
+* Items are added to the *back* of the queue when we
+  *enqueue* a new item.
+
+```mermaid
+flowchart
+  subgraph Queue-Actions
+    direction LR
+    New[NewLast/NewNewest] -- "(enqueue)" --> Last[Last/Newest]
+    subgraph Queue
+      direction LR
+      Last[Last/Newest] --> NextN[NextN] -- ... --> Next[Next] --> First[First/Oldest]
+    end
+    First[First/Oldest] -- "(dequeue)" --> Out["."]
+  end
+```
+
+So, you might be wonder: *Great! So what is the syntax for Elixir's
+`Queue` data structure?* There is no built-in Queue data structure;
+and this is why a Queue is called an "abstract data structure": You
+must implement the concept yourself.
+
+The most common implementation of a Queue is to base it on a List.
+
+### Your Turn
+
+How would you create a new empty Queue?
+
+```elixir
+defmodule Queue do
+  def new() do
+    :your_code_here
+  end
+end
+```
+
+Continuing, a new Queue should be empty of course!
+
+```elixir
+defmodule Queue do
+  def new() do
+    :your_code_here
+  end
+
+  def empty?(queue) do
+    :your_code_here
+  end
+end
+```
+
+Continuing, how would you *enqueue* an item -- let's say `:bob` -- to an empty Queue?
+
+```elixir
+defmodule Queue do
+  def new() do
+    :your_code_here
+  end
+
+  def empty?(queue) do
+    :your_code_here
+  end
+
+  def enqueue(queue, item) do
+    :your_code_here
+  end
+end
+```
+
+Continuing, it might be nice to know how many items are in the Queue.
+
+```elixir
+defmodule Queue do
```

Comment:
What do you think about having a single documented module for `Queue`? I think it's unclear having multiple of the same `Queue` module. We can also get weird evaluation issues when we have the same module in multiple cells.

Reply:
the goal here was to "build up" the Queue module. do you have an idea or example in a previous lesson i could templatize or model on?

Reply:
I like the build-up. What about making these code snippets in markdown to explain, then have the exercise in a single module at the end in a code cell to complete?

Reply:
It doesn't have the same build-up, but if you're looking for a good exercise example, then perhaps `CustomEnum with Reduce`.

Alternatively, how about we share this livebook (with no changes, keep the multiple cells) with @BigSpaces and @tindrew so they can give the exercise a try?